compiler/aarch64/codegen/OMRMachine.cpp

```diff
@@ -79,18 +82,380 @@ TR::RealRegister *OMR::ARM64::Machine::freeBestRegister(TR::Instruction *current
+            location = self()->cg()->allocateSpill(8, false, NULL);
+            }
+         break;
+      default:
+         TR_ASSERT(false, "Unsupported RegisterKind.");
+         break;
+      }
+   candidates[0]->setBackingStorage(location);
 
-   return NULL;
+   tmemref = new (self()->cg()->trHeapMemory()) TR::MemoryReference(currentNode, location->getSymbolReference(), 8, self()->cg());
+
+   if (!comp->getOption(TR_DisableOOL))
+      {
+      if (!self()->cg()->isOutOfLineColdPath())
+         {
+         TR_Debug *debugObj = self()->cg()->getDebug();
+         // the spilledRegisterList contains all registers that are spilled before entering
+         // the OOL cold path, post dependencies will be generated using this list
+         self()->cg()->getSpilledRegisterList()->push_front(candidates[0]);
+
+         // OOL cold path: depth = 3, hot path: depth = 2,  main line: depth = 1
+         // if the spill is outside of the OOL cold/hot path, we need to protect the spill slot
+         // if we reverse spill this register inside the OOL cold/hot path
+         if (!self()->cg()->isOutOfLineHotPath())
+            {// main line
+            location->setMaxSpillDepth(1);
+            }
+         else
+            {
+            // hot path
+            // do not overwrite main line spill depth
+            if (location->getMaxSpillDepth() != 1)
+               {
+               location->setMaxSpillDepth(2);
+               }
+            }
+         if (debugObj)
+            self()->cg()->traceRegisterAssignment("OOL: adding %s to the spilledRegisterList, maxSpillDepth = %d ",
+                                          debugObj->getName(candidates[0]), location->getMaxSpillDepth());
+         }
+      else
+         {
+         // do not overwrite mainline and hot path spill depth
+         // if this spill is inside OOL cold path, we do not need to protecting the spill slot
+         // because the post condition at OOL entry does not expect this register to be spilled
+         if (location->getMaxSpillDepth() != 1 &&
+             location->getMaxSpillDepth() != 2 )
+            {
+            location->setMaxSpillDepth(3);
+            self()->cg()->traceRegisterAssignment("OOL: In OOL cold path, spilling %s not adding to spilledRegisterList", (candidates[0])->getRegisterName(self()->cg()->comp()));
+            }
+         }
+      }
+
+   if (self()->cg()->comp()->getOption(TR_TraceCG))
+      {
+      diagnostic("\n\tspilling %s (%s)",
+                  best->getAssignedRegister()->getRegisterName(self()->cg()->comp()),
+                  best->getRegisterName(self()->cg()->comp()));
+      }
+
+   switch (rk)
+      {
+      case TR_GPR:
+         generateTrg1MemInstruction(self()->cg(), TR::InstOpCode::ldrimmx, currentNode, best, tmemref, currentInstruction);
+         break;
+      case TR_FPR:
+         generateTrg1MemInstruction(self()->cg(), TR::InstOpCode::vldrimmd, currentNode, best, tmemref, currentInstruction);
```

Comment:
These cases only differ on the instruction opcode.  Can they be combined to save some code space?

Reply:
Merged the GPR and FPR cases.